### ISSUE TITLE
alexanderism55: Behavioural, strategy implmentation, 3 weeks

### ISSUE THREAD

**Issue by alexanderism55** [lab-brussels-1/home#362]
> [my fork](https://github.com/alihaidermalik20/solution-write-ups)

## Learning Objectives

<details>
<summary>Priorities: 🥚, 🐣, 🐥, 🐔 (click to learn more)</summary>
<br>

There is a lot to learn in this repository. If you can't master all the material
at once, that's expected! Anything you don't master now will always be waiting
for you to review when you need it. These 4 emoji's will help you prioritize
your study time and to measure your progress:

- 🥚: Understanding this material is required, it covers the base skills you'll
  need to move on. You do not need to finish all of them but should feel
  comfortable that you could with enough time.
- 🐣: You have started all of these exercises and feel you could complete them
  all if you just had more time. It may not be easy for you but with effort you
  can make it through.
- 🐥: You have studied the examples and started some exercises if you had time.
  You should have a big-picture understanding of these concepts/skills, but may
  not be confident completing the exercises.
- 🐔: These concepts or skills are not necessary but are related to this module.
  If you are finished with 🥚, 🐣 and 🐥 you can use the 🐔 exercises to push
  yourself without getting distracted from the module's main objectives.

---

</details>

### [1. Remix](./1-remix/)

Practice studying and remixing other people's solutions to coding challenges. Create your own solutions by mixing and matching pieces from other people's code.

- 🐣 **Reconstructing**: You can reconstruct a variety of solutions to the same challenge when they are presented as parsons problems.
- 🐣 **Analyzing**: You can analyze a function written at your level. This includes:
  - [x] **Behavior**: Write documentation, test cases and use cases to describe the function's behavior.
  - [x] **Strategy**: Describe the function's strategy using plain english.
  - [x] **Implementation**: List the language features in a function and explain how each one is used.
  - [x] **Small Changes**: You can think of 2+ changes to the function's _implementation_ that would not change it's _strategy_.
- 🐣 **Remixing** You can analyze several solutions to the same challenge then ...
  - [x] **Write**: Your own solution by remixing the ones you studied.
  - [x] **Explain**: How the other solutions inspired yours; what ideas did you take from them? what ideas did you _not_ take?
  - [x] **Analyze**: Complete a write-up your own solution as though someone else wrote it.

### [2. Write](./2-write/)

- 🐣 **Function Design**:
  - [x] **Writing Tests**: Given a working function, you can write passing test cases to describe it's behavior.
  - [ ] **Writing Functions**: You can design multiple solutions to the same code challenge, keeping notes about different experiments you tried along the way.
- [ ] 🐣 **Generating Documentation**: You can write a JSDoc comment for your solutions and run a script to generate markdown documentation.
- [ ] 🐣 **Fuzz Testing**: You write solutions that pass randomly generated test cases.
- 🐣 **Test Driven Development**: You can solve open-ended, ambiguous coding challenges at your level:
  - [x] **Reading Docs**: You can understand what the function is supposed to do by reading it's JSDoc description.
  - [x] **Writing Tests**: You can write test cases _before_ there is a function to test.
  - [x] **Writing Functions**: You can write one function that passes the test cases you have prepared (even if it's just 1 test case!).
  - [x] **Refactoring**: You can improve your function's implementation _without_ failing any test cases that were passing.
  - [x] **Iterative Development**: You can repeat the TDD process until you are satisfied with your test cases and solution.
- [ ] 🐔 **Code Golf**: Write your solutions with the fewest characters possible! This won't help you write readable code, but it will make you think deeply about JS, your strategy and implementation.

### [3. Review](./3-review/)

- 🐣 **Continuous Integration**: You can check your code's quality _before_ pushing so your CI checks all pass.
  - [ ] **Formatting**
  - [ ] **Linting**
  - [ ] **Testing**
- [ ] 🐣 **Code Review**: You can use a checklist to give a thorough, positive and constructive review of your classmates' solutions.
- [ ] 🐔 **Code Coverage**: You can explain what code coverage is, why it's important, and can write unit tests with 100% code coverage.

---

**Comment by alexanderism55** [comment]
Week 1


## I Need Help With:
- sharpening my testing skills. I use test instead of description, and i dont use an it statement. However i use equal and expected.
## What went well?
Understanding Jest was quite easy. The entire idea of testing was far more easier than what my fears can justify. In fact testing helped me write my function better, or shall i say it helped me think how to write better function. 
## What went less well?
I do not know why but I couldn't get the 'description', and the 'it' function rolling. Also i coudnt export my function using 'module.exports = code ' and import using 'code = require './location'. When I did that, it said that require was not defined. So it gave me an option to press change to "enp or ens" i m not sure but when i pressed it it transforms it self to 'import code from location'.
## Lessons Learned
- I learned that testing can help you be a more professional coder. What i mean by that is that if I am going to build a good program; the way I handle the project will matter ten times more than how good I can write a function. So in a way, I understand what you guys said at the beginning of HYF, we wont just teach u how to code, but how to be a good team member. I also learned to write functions better. I mean I wrote a function to turn hours and minutes to seconds, in about ten minutes. Two weeks ago I did not even know how to console.log properly.

## Sunday Prep Work
- I will try a challenge a from the "edbit" website and write test for it.

**Comment by danielhalasz** [comment]
> * I use test instead of description, and i dont use an it statement. However i use equal and expected.

I hope this has become more clear yesterday..but all tests are supposed to follow the logic of describe>it/test>expect

**Comment by danielhalasz** [comment]
> Also i coudnt export my function using 'module.exports = code ' and import using 'code = require './location'. When I did that, it said that require was not defined.

you can use import instead of require, as it was in all the repo examples

**Comment by danielhalasz** [comment]
> I do not know why but I couldn't get the 'description', and the 'it' function rolling.

it is just supposed to be a description of what you are doing, a bit like a comment

**Comment by alexanderism55** [comment]
Week 1


## I Need Help With:
- Nothing so far everything is going super smooth. I did 424 tests which super cool.
## What went well?
- Using jest went really well, a little tutorial helped me understand that there is a lot of cool features with jest. 
## What went less well?
- Syntax for the test is very easy to break and can get really confusing, i broke the test so many times.
## Lessons Learned
- Writing tests makes you fell confident when upgrading your function and it also shows where your function is lacking, my function was lacking a bit which testing helped me fix.

## Sunday Prep Work
- Im gonna get prepared for it today.

**Comment by danielhalasz** [comment]
> * Nothing so far everything is going super smooth. I did 424 tests which super cool.

wow, nice job Alex 💪

**Comment by danielhalasz** [comment]
> * Syntax for the test is very easy to break and can get really confusing, i broke the test so many times.

yes, I feel your pain..many brackets and curly braces 🙂 it is good practice though! you will be a master spotter
